File: medplum-java/medplum-model/src/main/java/com/fhir/AllergyintoleranceCategoryItem.java
package com.fhir;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import java.lang.Object;
import java.lang.Override;
import java.lang.String;
import java.util.Locale;
import javax.annotation.Nonnull;

public final class AllergyintoleranceCategoryItem {
  public static final AllergyintoleranceCategoryItem ENVIRONMENT = new AllergyintoleranceCategoryItem(Value.ENVIRONMENT, "ENVIRONMENT");

  public static final AllergyintoleranceCategoryItem FOOD = new AllergyintoleranceCategoryItem(Value.FOOD, "FOOD");

  public static final AllergyintoleranceCategoryItem BIOLOGIC = new AllergyintoleranceCategoryItem(Value.BIOLOGIC, "BIOLOGIC");

  public static final AllergyintoleranceCategoryItem MEDICATION = new AllergyintoleranceCategoryItem(Value.MEDICATION, "MEDICATION");

  private final Value value;

  private final String string;

  AllergyintoleranceCategoryItem(Value value, String string) {
    this.value = value;
    this.string = string;
  }

  public Value getEnumValue() {
    return value;
  }

  @Override
  @JsonValue
  public String toString() {
    return this.string;
  }

  @Override
  public boolean equals(Object other) {
    return (this == other) 
      || (other instanceof AllergyintoleranceCategoryItem && this.string.equals(((AllergyintoleranceCategoryItem) other).string));
  }

  @Override
  public int hashCode() {
    return this.string.hashCode();
  }

  public <T> T accept(Visitor<T> visitor) {
    switch (value) {
      case ENVIRONMENT:
        return visitor.visitENVIRONMENT();
      case FOOD:
        return visitor.visitFOOD();
      case BIOLOGIC:
        return visitor.visitBIOLOGIC();
      case MEDICATION:
        return visitor.visitMEDICATION();
      case UNKNOWN:
      default:
        return visitor.visitUnknown(string);
    }
  }

  @JsonCreator(
      mode = JsonCreator.Mode.DELEGATING
  )
  public static AllergyintoleranceCategoryItem valueOf(@Nonnull String value) {
    String upperCasedValue = value.toUpperCase(Locale.ROOT);
    switch (upperCasedValue) {
      case "ENVIRONMENT":
        return ENVIRONMENT;
      case "FOOD":
        return FOOD;
      case "BIOLOGIC":
        return BIOLOGIC;
      case "MEDICATION":
        return MEDICATION;
      default:
        return new AllergyintoleranceCategoryItem(Value.UNKNOWN, upperCasedValue);
    }
  }

  public enum Value {
    FOOD,

    MEDICATION,

    ENVIRONMENT,

    BIOLOGIC,

    UNKNOWN
  }

  public interface Visitor<T> {
    T visitFOOD();

    T visitMEDICATION();

    T visitENVIRONMENT();

    T visitBIOLOGIC();

    T visitUnknown(String unknownType);
  }
}

File: medplum-typescript/src/model/fhir/AllergyintoleranceCategoryItem.ts
export type AllergyintoleranceCategoryItem =
    | AllergyintoleranceCategoryItem.Food
    | AllergyintoleranceCategoryItem.Medication
    | AllergyintoleranceCategoryItem.Environment
    | AllergyintoleranceCategoryItem.Biologic;

export const AllergyintoleranceCategoryItem = {
    Food: "food" as AllergyintoleranceCategoryItem.Food,
    Medication: "medication" as AllergyintoleranceCategoryItem.Medication,
    Environment: "environment" as AllergyintoleranceCategoryItem.Environment,
    Biologic: "biologic" as AllergyintoleranceCategoryItem.Biologic,

    _visit: <Result>(value: AllergyintoleranceCategoryItem, visitor: AllergyintoleranceCategoryItem._Visitor<Result>): Result => {
        switch (value) {
            case AllergyintoleranceCategoryItem.Food: return visitor.food();
            case AllergyintoleranceCategoryItem.Medication: return visitor.medication();
            case AllergyintoleranceCategoryItem.Environment: return visitor.environment();
            case AllergyintoleranceCategoryItem.Biologic: return visitor.biologic();
            default: return visitor._unknown();
        }
    },
};

export declare namespace AllergyintoleranceCategoryItem {
    export type Food = "food" & {
        __Food: void,
    };
    export type Medication = "medication" & {
        __Medication: void,
    };
    export type Environment = "environment" & {
        __Environment: void,
    };
    export type Biologic = "biologic" & {
        __Biologic: void,
    };

    export interface _Visitor<Result> {
        food: () => Result;
        medication: () => Result;
        environment: () => Result;
        biologic: () => Result;
        _unknown: () => Result;
    }
}
